scripts-components/package.json

```diff
@@ -0,0 +1,5 @@
+{
+  "devDependencies": {
+    "@mangoweb/scripts-base": "^0.0.10"
```

Comment:
Ta stříška by tu být neměla. Major `0` povoluje breaking changes kdykoliv. I při přechodu z `0.0.10` na `0.0.11`. https://semver.org/#spec-item-4

Reply:
To je schválně ‒ bez stříšky by to měly mít samotné komponenty, ale tady je to podle mě takhle dobře. Tohle není nainstalovatelný balíček, ale je jen způsob, jak pro vývoj mít ve všech komponentách v tomto adresáři nainstalované `scripts-base`, když je to společná závislost. Když `scripts-base` udělají zpětně nekompatibilní změnu, naopak chceme, aby se to ve vývojovém prostředí rozbilo a někdo na to přišel. Na ty, kdo instalují ty komponenty, tohle ale nemá vliv.

Reply:
Aha. Už to chápu.

Souhlasím s tebou, že peerDependencies, nefungují, jak by se nám hodilo, pokud jsem tě správně pochopil. Rozhodně v nich vidíme oba problém. Tyhle dependencies o level výše vše IMHO ještě více zesložiťují. Dříve, když jsem si zapomněl nainstaloval peerDependencies, tak mi failnul publish, při kterém se buildem testuje, že je kód ok. Teď se fallbackne na `node_modules` o úroveň výš, což je trochu nezvyklé. Ve `script-components` může být jiná verze `scripts-base` než čekám. Publish mi třeba projde, ale publishnutá verze bude v `peerDependencies` odkazovat na verzi `scripts-base`, která třeba kompatibilní není.

Reply:
Co když `scripts-base` není ani `devDependency` ani `peerDependency`, ale prostě `dependency`? Vždyť takové `InView` nebo `Shapes` na tom přímo závisí, protože extendují `Component`, což se netýká jen vývoje, ale i produkce.

Reply:
Problém je dost možná v tom, že jsme stále v major `0`, takže se v dependencies těžko udržují vazby postavené na semveru.

Reply:
Myslím, že `peerDependencies` chceme, abychom vynutili, že vždy bude v bundlu přítomna maximálně jedna instance té závislosti. Podle kódu to je asi fakticky normální `dependency`, ale takhle jsme si na sebe záměrně ušili bič, abychom prostě pohlídali, že ten bundle nebude bobtnat. Máme pod kontrolou obě strany: jak tuto knihovnu, tak projekty, tak toho podle mě rozumně jde docílit. Jako vždy, UX stojí nad DX, takže menší bundle je prostě win.

Ty `peerDependencies` lze řešit pomocí `npm install --no-save @mangoweb/scripts-base` v komponentě, ale to podle mě v praxi nic moc neřeší ‒ i tady se snadno může rozejít verze deklarovaná v `package.json` s tou, kterou si člověk takhle nainstaluje. Jediný rozdíl je, že to člověk najednou musí řešit ve všech komponentách, což je režijně nepatrně složtější. Jinak to trpí úplně všemi neduhy, které popisuješ. I tady může být _jiná verze, než čekáš_. I tady si člověk může vesele nainstalovat novější `scripts-base` a pak zapomenout na update peer dependency v `package.json` Co tedy reálně navrhuješ? Mít místo jedné instance problému těch problémů x?

Mohou tu nastat jen tři scénáře: verze v adresáři `scropts-components` je 1) novější, 2) stejná nebo 3) starší, než má v `peerDependencies` komponenta. V praxi ovšem bude typicky nejnovější vydaná, tedy 1) nebo 2) ‒ vydávám novou verzi `scripts-base`, tak bumpnu i tu v tom adresáři. Pokud při vydávání nové verze zapomenu změnit tento adresář, není to problém, pokud to nebyl problém již předtím.

- **Případ 1):** Pokud mi teď projde build a komponenta funguje, změny byly zpětně kompatibilní a není žádný problém. Pokud build neprojde, je třeba to opravit, a je tudíž super, že mě něco donutilo to řešit. Za mě v pořádku stav. Samozřejmě může projít build, ale fungovat to nebude, ale to už je na tom, kdo publikuje novou verzi komponenty. Stejně se očekává se, že si to otestuje. Stále může zapomenout aktualizovat tu peer dependency, ale jak jsem zmínil, to je prostě nevyřešenej problém tak či onak.
- **Případ 2):** To je ideální stav, není žádný problém.
- **Případ 3):** Tohle by nebyl moc dobrý stav, ale nemyslím si, že k němu v praxi kdy dojde. Reálně by se to mohlo stát, pouze pokud by někdo vydával novou verzi komponenty, která závisí na novější verzi `scripts-base`. Když si to bude testovat tak, že `scripts-base` bude brát odsud, k tomuhle z podstaty věci nedojde.

Proč je tedy tohle špatně? K čemu přesně jak dojde, že tohel bude o tolik horší, že si kvůli tomu každej budeme managovat x verzí `scripts-base`?

Reply:
Mé i tvé řešení stojí na tom, že moc dobře víme, jak je to seskládané dohromady a navzájem se hlídáme. Přijde mi, že tvé řešení sice nic nezhoršuje, ale ani nezlepšuje. Osobně bych se do něj tedy nepouštěl, ale o tom už není řeč, když je to hotové. :)

Klidně klikni na resolve. Tvé řešení podporuji. :+1:

Problém je, že si musíme dávat pozor, co za updaty děláme a znesnadňujeme onboarding dalších vývojářů. Napadlo mne, jestli není částečná cesta ven přejít už na major verze. Snížilo by se tím riziko, že by se do bundlu dostalo dvoje `scripts-base` i přes to, že by jednotlivé komponenty měly `scripts-base` jako `dependency`.